server/src/copilot/api/llmModelManager.ts

```diff
@@ -0,0 +1,129 @@
+import { generateContent as generateGeminiContent, generateEmbeddings as generateGeminiEmbeddings } from './gemini';
+import { generateContent as generateOpenAIContent, generateEmbeddings as generateOpenAIEmbeddings } from './openai';
+import { generateContent as generateAnthropicContent } from './anthropic';
+import { generateContent as generateHuggingfaceContent } from './huggingface';
+import { getPromptFromCache, setPromptToCache } from '../utils/promptCache';
+import { incorporateSuggestion } from '../utils/preparePrompt';
+import { REGEX, PROVIDERS, DEFAULTS } from '../utils/constants';
+import { generateCacheKey } from '../utils/cacheKeyGenerator';
+import { beautifyConcertoCode, cleanSuggestion } from '../utils/responseProcessor';
+import { agentPlanner } from '../agent/agentPlanner';
+import { ModelConfig, DocumentDetails, PromptConfig, Documents, Embedding } from '../utils/types';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { GLOBAL_STATE } from '../../state';
+import { Lock } from '../utils/lock';
+import { log } from '../../state';
+import { handleConcertoDocumentChange } from '../../documents/concertoHandler';
+
+const lock = new Lock();
+
+async function generateContentByProvider(provider: string, config: ModelConfig, promptArray: Array<{ content: string, role: string }>): Promise<string> {
+    switch (provider) {
+        case PROVIDERS.GEMINI:
+            return await generateGeminiContent(config, promptArray);
+        case PROVIDERS.OPENAI:
+            return await generateOpenAIContent(config, promptArray);
+        case PROVIDERS.ANTHROPIC:
+            return await generateAnthropicContent(config, promptArray);
+        case PROVIDERS.HUGGINGFACE:
+            return await generateHuggingfaceContent(config, promptArray);
+        default:
+            throw new Error('Invalid model name specified in config');
+    }
+}
+
+export async function generateEmbeddingsByProvider(provider: string, config: ModelConfig, text: string): Promise<Embedding[]> {
+    switch (provider) {
+        case PROVIDERS.GEMINI:
+            return await generateGeminiEmbeddings(config, text);
```

Comment:
I think the name may also be a little confusing - Model Manager is a well established name in AP!

E.g. something like:

interface LargeLanguageModel  {
   getIdentifier() // id / key for the LLM
 ... whatever methods you need for an LLM (invoke/config etc)
}

class LargeLanguageModelProvider {
 - register(LargeLanguageModel)
 - unregister(id)
 - get(id)
} 

This will make it easier for folks to understand what interface they need to implement to plugin a new LLM, and how to register it with the system.

Reply:
Done: 
1. Refactored the code to introduce an LargeLanguageModel interface that defines the common methods (getIdentifier, generateContent, generateEmbeddings) required by all LLM providers. This is looking more clean now. 
2. Renamed the llmModelManager to llmManager.